---------------------------------------------------------------------

# Analysez les ventes de l'Entreprise "Rester livres"

Étude d'Analyse de données de vente projet_4 -Ikhlas(Novembre 2021)


Je suis Data Analyst d'une grande chaîne de librairie, fraîchement embauché depuis une semaine ! "Rester livres" s'est d'abord développée dans une grande ville de France, avec plusieurs magasins, jusqu'à décider d'ouvrir une boutique en ligne. Son approche de la vente de livres en ligne, basée sur des algorithmes de recommandation, lui a valu un franc succès !


In [1]:
# system libraries
import warnings
warnings.filterwarnings('ignore')
warnings.simplefilter(action='ignore', category=FutureWarning)


#Import des librairies Python

import pandas as pd
import numpy as np

import datetime as dt


Une fois les différents modules importés, on charge les fichiers bruts nécessaires à notre analyse. On dispose d'un jeu de données composé de trois fichiers : un fichier contenant les ventes, nommé transactions, un fichier contenant les informations concernant les clients nommé customers et un fichier contenant les informations les produits vendus se nommant products 

In [2]:
#Import des fichiers csv extraits de la base de données de l'entreprise
client = pd.read_csv('customers.csv')#la liste des clients
produit = pd.read_csv('products.csv')#la liste des produits
transaction = pd.read_csv('transactions.csv')#les ventes (appelées “Transactions”)

# Mission_1

# Nettoyage des données

on affiche les premières lignes des fichiers  pour voir leur composition

In [5]:
transaction.head()

id_prod                        date session_id client_id
0  0_1483  2021-04-10 18:37:28.723910    s_18746    c_4450
1   2_226  2022-02-03 01:55:53.276402   s_159142     c_277
2   1_374  2021-09-23 15:13:46.938559    s_94290    c_4270
3  0_2186  2021-10-17 03:27:18.783634   s_105936    c_4597
4  0_1351  2021-07-17 20:34:25.800563    s_63642    c_1242

In [6]:
produit.head()

id_prod  price  categ
0  0_1421  19.99      0
1  0_1368   5.13      0
2   0_731  17.99      0
3   1_587   4.99      1
4  0_1507   3.99      0

In [7]:
client.head()

client_id sex  birth
0    c_4410   f   1967
1    c_7839   f   1975
2    c_1699   f   1984
3    c_5961   f   1962
4    c_5320   m   1943

In [8]:
# merger les 2 tableaux: produit & transaction 
df_transactions_products = pd.merge(produit, transaction, left_on='id_prod', right_on='id_prod')
df_transactions_products

id_prod  price  categ                        date session_id client_id
0       0_1421  19.99      0  2021-05-12 23:26:29.885083    s_33709    c_7954
1       0_1421  19.99      0  2022-01-25 11:27:39.357012   s_154957     c_254
2       0_1421  19.99      0  2022-01-02 22:29:33.796003   s_144068    c_5770
3       0_1421  19.99      0  2021-03-05 17:10:40.265676     s_2187    c_1455
4       0_1421  19.99      0  2021-03-29 13:55:04.422637    s_13112    c_2768
...        ...    ...    ...                         ...        ...       ...
336908   1_140  38.56      1  2021-03-30 11:10:45.544437    s_13465     c_974
336909   1_140  38.56      1  2021-12-25 06:33:13.259886   s_139663    c_4766
336910   1_140  38.56      1  2022-01-01 02:58:25.933977   s_143171    c_6279
336911  0_1920  25.16      0  2021-04-13 18:36:10.252971    s_20115    c_7088
336912  0_1920  25.16      0  2021-05-30 02:37:22.371278    s_41465    c_7748

[336913 rows x 6 columns]

In [10]:
# merger les 2 tableaux: produit & transaction et client
df_transactions_products_client= pd.merge(client,df_transactions_products, left_on='client_id', right_on='client_id')
df_transactions_products_client

client_id sex  birth id_prod  price  categ                        date  \
0         c_4410   f   1967   1_385  25.99      1  2021-03-22 01:40:22.782925   
1         c_4410   f   1967  0_1110   4.71      0  2021-11-04 16:28:30.169021   
2         c_4410   f   1967  0_1111  19.99      0  2021-03-22 01:27:49.480137   
3         c_4410   f   1967   1_461  12.99      1  2021-08-11 08:40:47.495793   
4         c_4410   f   1967   1_536  11.21      1  2022-01-18 17:05:07.468131   
...          ...  ..    ...     ...    ...    ...                         ...   
336908      c_84   f   1982  0_1452  13.77      0  2021-09-20 10:11:35.510099   
336909      c_84   f   1982   1_158  22.17      1  2021-10-01 16:11:20.841361   
336910      c_84   f   1982  0_1438   9.31      0  2021-12-29 06:11:50.316631   
336911      c_84   f   1982   1_459  15.99      1  2022-02-17 00:16:56.629536   
336912      c_84   f   1982   1_110  54.90      1  2021-09-13 21:39:46.845376   

       session_id  
0          s_9707  
1        s_114715  
2          s_9707  
3         s_74236  
4        s_151740  
...           ...  
336908    s_92701  
336909    s_98493  
336910   s_141701  
336911   s_166337  
336912    s_89478  

[336913 rows x 8 columns]

In [14]:
#Analyse des modalités de la variable 'price' 
#Fonction .unique() permet de les afficher
df_transactions_products_client['price'].unique()

array([25.99,  4.71, 19.99, ..., 20.26, 31.98, 21.82])

In [15]:
#regroupement de données price par rapport id_prod
df_transactions_products_client.groupby('id_prod')['price'].unique()

id_prod
0_0         [3.75]
0_1        [10.99]
0_10       [17.95]
0_100       [20.6]
0_1000      [6.84]
            ...   
2_96       [47.91]
2_97      [160.99]
2_98      [149.74]
2_99       [84.99]
T_0         [-1.0]
Name: price, Length: 3265, dtype: object

In [16]:
#regroupement de données price par rapport session_id
df_transactions_products_client.groupby('session_id')['price'].unique()

session_id
s_0                             [-1.0]
s_1                            [11.99]
s_10                           [26.99]
s_100                   [13.73, 19.99]
s_1000       [18.93, 4.99, 9.31, 5.99]
                      ...             
s_99994            [5.48, 13.99, 9.45]
s_99995                        [19.84]
s_99996    [13.99, 18.99, 11.99, 11.3]
s_99997                         [6.99]
s_99998            [8.99, 8.61, 17.51]
Name: price, Length: 169174, dtype: object

## Analyse des données manquantes

In [17]:
# Dénombrement des valeurs manquantes
df_transactions_products_client.isna().sum()

client_id     0
sex           0
birth         0
id_prod       0
price         0
categ         0
date          0
session_id    0
dtype: int64

Le fichier  ne présentent pas de données manquantes

## Analyse des données dupliquées

In [18]:
df_transactions_products_client.duplicated('client_id').sum()

328311

On crée une variable qui contient les données dupliquées. Ensuite, on affiche les données dupliquées du DataFrame.

In [19]:
# création d'une variable contenant les informations dupliquées
df2 = df_transactions_products_client[df_transactions_products_client.duplicated()]


In [20]:
df2

client_id sex  birth id_prod  price  categ  \
108163      ct_0   f   2001     T_0   -1.0      0   
108170      ct_0   f   2001     T_0   -1.0      0   
108173      ct_0   f   2001     T_0   -1.0      0   
108174      ct_0   f   2001     T_0   -1.0      0   
108177      ct_0   f   2001     T_0   -1.0      0   
...          ...  ..    ...     ...    ...    ...   
332677      ct_1   m   2001     T_0   -1.0      0   
332678      ct_1   m   2001     T_0   -1.0      0   
332679      ct_1   m   2001     T_0   -1.0      0   
332680      ct_1   m   2001     T_0   -1.0      0   
332681      ct_1   m   2001     T_0   -1.0      0   

                                   date session_id  
108163  test_2021-03-01 02:30:02.237443        s_0  
108170  test_2021-03-01 02:30:02.237429        s_0  
108173  test_2021-03-01 02:30:02.237446        s_0  
108174  test_2021-03-01 02:30:02.237446        s_0  
108177  test_2021-03-01 02:30:02.237442        s_0  
...                                 ...        ...  
332677  test_2021-03-01 02:30:02.237416        s_0  
332678  test_2021-03-01 02:30:02.237425        s_0  
332679  test_2021-03-01 02:30:02.237423        s_0  
332680  test_2021-03-01 02:30:02.237421        s_0  
332681  test_2021-03-01 02:30:02.237431        s_0  

[126 rows x 8 columns]

On observe la présence de 126 lignes correspondant à des sessions de tests. Il faudra donc les supprimer du fichier. 

On supprime donc les valeurs dupliquées correspondant aux sessions test du 1er mars 2021.

In [21]:
df_transactions_products_client = df_transactions_products_client.drop_duplicates()
df_transactions_products_client[df_transactions_products_client.duplicated()]

Empty DataFrame
Columns: [client_id, sex, birth, id_prod, price, categ, date, session_id]
Index: []

Les valeurs dupliquées sont donc supprimées mais il reste encore des données de session test avec l'identifiant s_0. On va donc supprimer ces valeurs pour pouvoir faire son analyse.

In [22]:
print(df2['id_prod'].unique())

['T_0']


In [23]:
print(df2['client_id'].unique())

['ct_0' 'ct_1']


In [24]:
print(df2['date'].unique())


['test_2021-03-01 02:30:02.237443' 'test_2021-03-01 02:30:02.237429'
 'test_2021-03-01 02:30:02.237446' 'test_2021-03-01 02:30:02.237442'
 'test_2021-03-01 02:30:02.237436' 'test_2021-03-01 02:30:02.237413'
 'test_2021-03-01 02:30:02.237435' 'test_2021-03-01 02:30:02.237424'
 'test_2021-03-01 02:30:02.237428' 'test_2021-03-01 02:30:02.237419'
 'test_2021-03-01 02:30:02.237437' 'test_2021-03-01 02:30:02.237441'
 'test_2021-03-01 02:30:02.237414' 'test_2021-03-01 02:30:02.237432'
 'test_2021-03-01 02:30:02.237430' 'test_2021-03-01 02:30:02.237427'
 'test_2021-03-01 02:30:02.237426' 'test_2021-03-01 02:30:02.237449'
 'test_2021-03-01 02:30:02.237417' 'test_2021-03-01 02:30:02.237415'
 'test_2021-03-01 02:30:02.237416' 'test_2021-03-01 02:30:02.237420'
 'test_2021-03-01 02:30:02.237444' 'test_2021-03-01 02:30:02.237440'
 'test_2021-03-01 02:30:02.237423' 'test_2021-03-01 02:30:02.237438'
 'test_2021-03-01 02:30:02.237445' 'test_2021-03-01 02:30:02.237412'
 'test_2021-03-01 02:30:02.237439'

Pour supprimer toutes les données issues des tests, on supprime toutes les sessions dont l'identifiant session vaut 's_0'. On crée donc une liste contenant les index des lignes avec l'identifiant de session s_0 puis on supprime les lignes associées à l'aide de la méthode .drop().

In [25]:
#création d'une liste contenant les index avec l'identifiant de session s_0
index_s_0 = df_transactions_products_client[df_transactions_products_client['session_id'] == 's_0'].index.values
df_transactions_products_client.drop(index_s_0, 0, inplace=True)

In [26]:
df_transactions_products_client.loc[df_transactions_products_client['session_id'] == 's_0']

Empty DataFrame
Columns: [client_id, sex, birth, id_prod, price, categ, date, session_id]
Index: []

In [27]:
#taille
client.shape

(8623, 3)

In [28]:
transaction.shape

(337016, 4)

In [29]:
df_transactions_products_client.shape

(336713, 8)

## Analyse des types

In [30]:
print(df_transactions_products_client.dtypes)

client_id      object
sex            object
birth           int64
id_prod        object
price         float64
categ           int64
date           object
session_id     object
dtype: object


On remarque que toutes les variables sont considérées comme des objets. On peut transformer la variable 'date' en type 'datetime'.

La variable 'sex' apparaît comme un objet, il peut être plus intéressant de changer ce type en variable catégorielle.

In [31]:
df_transactions_products_client['date'] = pd.to_datetime(df_transactions_products_client['date'])
df_transactions_products_client['sex'] = df_transactions_products_client['sex'].astype('category')

In [32]:
#Il serait plus intéressant de modifier le type de la colonne 'categ' en une valeur catégorielle.
df_transactions_products_client['categ'] = df_transactions_products_client['categ'].astype('category')
df_transactions_products_client.dtypes

client_id             object
sex                 category
birth                  int64
id_prod               object
price                float64
categ               category
date          datetime64[ns]
session_id            object
dtype: object

In [33]:
#Une fois cette transformation faite, il est plus simple d'organiser les données par ordre chronologique.
df_transactions_products_client = df_transactions_products_client.sort_values('date')
df_transactions_products_client = df_transactions_products_client.set_index('date')
print(df_transactions_products_client.head())

                           client_id sex  birth id_prod  price categ  \
date                                                                   
2021-03-01 00:01:07.843138     c_329   f   1967  0_1259  11.99     0   
2021-03-01 00:02:26.047414     c_664   m   1960  0_1390  19.37     0   
2021-03-01 00:02:38.311413     c_580   m   1988  0_1352   4.50     0   
2021-03-01 00:04:54.559692    c_7912   f   1989  0_1458   6.55     0   
2021-03-01 00:05:18.801198    c_2033   f   1956  0_1358  16.49     0   

                           session_id  
date                                   
2021-03-01 00:01:07.843138        s_1  
2021-03-01 00:02:26.047414        s_2  
2021-03-01 00:02:38.311413        s_3  
2021-03-01 00:04:54.559692        s_4  
2021-03-01 00:05:18.801198        s_5  


In [34]:
#On peut ainsi classer les produits par ordre de catégorie.
df_transactions_products_client= df_transactions_products_client.sort_values('categ')
print(df_transactions_products_client)

                           client_id sex  birth id_prod  price categ  \
date                                                                   
2021-03-01 00:01:07.843138     c_329   f   1967  0_1259  11.99     0   
2021-09-30 13:32:32.789640    c_2298   m   1972   0_674  13.95     0   
2021-09-30 13:33:14.103844    c_7714   f   1972  0_2029   3.99     0   
2021-09-30 13:36:06.900756     c_441   f   1972  0_2291  12.99     0   
2021-09-30 13:36:08.951808    c_7714   f   1972   0_699  18.99     0   
...                              ...  ..    ...     ...    ...   ...   
2022-02-06 09:16:52.981557    c_7533   f   2004   2_184  49.61     2   
2021-08-30 15:55:35.071302    c_4381   m   2004   2_102  59.14     2   
2021-05-04 14:14:05.187463    c_5382   f   1995   2_164  52.26     2   
2021-11-13 04:13:13.607660    c_6528   m   2004   2_115  50.99     2   
2021-04-20 03:55:49.505401     c_330   m   1953   2_207  50.99     2   

                           session_id  
date                   

In [35]:
# Reset les index pour pas les dupliquer et rename_axis pour éviter que les noms de colonnes soient au dessus des autres.
df_transactions_products_client.rename_axis(None,axis=1)
df_transactions_products_client.reset_index(inplace= True)
df_transactions_products_client.head()

date client_id sex  birth id_prod  price categ  \
0 2021-03-01 00:01:07.843138     c_329   f   1967  0_1259  11.99     0   
1 2021-09-30 13:32:32.789640    c_2298   m   1972   0_674  13.95     0   
2 2021-09-30 13:33:14.103844    c_7714   f   1972  0_2029   3.99     0   
3 2021-09-30 13:36:06.900756     c_441   f   1972  0_2291  12.99     0   
4 2021-09-30 13:36:08.951808    c_7714   f   1972   0_699  18.99     0   

  session_id  
0        s_1  
1    s_97930  
2    s_97943  
3    s_97931  
4    s_97943

## Enregistrements de fichiers nettoyé

In [36]:
#Une fois le nettoyage des fichiers terminés, on les enregistre en nouveau fichier csv pour l'étape d'analyse qui suit cette partie
df_transactions_products_client.to_csv("transactions_products_client.csv",index = False)